### DISCUSSION TITLE
"You don't have permission to access this" when creating/accesing items in items.create hook

### DISCUSSION THREAD

**Discussion by younky-yang** [directus/directus#13503] [Archive]
I got a lot of messages of "You don't have permission to access this." in my custom hooks for item.Create. For example I have a collection and have registered a hook for this collection's Create event. If I create the collection item directly from its own page, the hook works perfectly as I can access the other collection and update other collections. But if I create this collection item from another collection item which has a O2M relations, the hook will throw the "You don't have permission to access this." error for me which all the operation in the hooks that need to access DB failed. 

I dig into the source codes and found that the Items.Create is emitted after the transaction with the same knex connection. My assumption is that the transaction is done which I should have the opportunity to operate on the newly created records in my hook. The reality is no, so I add a 50ms delay at the beginning of my hook, and then it works for my local environment, but failed on my AWS environment, then I updated the delay to 200ms, it works both on my local and AWS environment. 

I also did the check of the db connection and I found the observed value of the pg response time is with 5ms, and the threshold is 150, so I don't think the DB connection is a problem for me. 

So needs some help on this issue.

**Comment by rijkvanzanten** [comment]
What event are you using? What payload are you submitting in an update? What does your schema look like? Are you using `action` or `filter`? What does your hook implementation look like? 🙂

**Reply by younky-yang** [reply]
It is just an action hook that filter Items.Create event. The payload is right with the collection item info in both cases (created in its own page or from the referenced item page). Also I didn't notice the schema difference from these 2 cases.

Will do some test with another case today to check what happens.

**Comment by younky-yang** [comment]
Did some test and below is my analysis result. 

Let's setup the collection A and B, and A and B has M2M relations which there is another table created with name A_B, when I create A with a new Item B in the A's item detail page, it first trigger the action hook of B.Create, then A_B.Create, last A.Create.

My custom hook will handle B.Create, so when it is triggered I tried to call readOne() of A, then I will get the issue of "You don't have permission to access this". I think the reason is the hook is called in async mode, and at that time, perhaps the transaction of updating A is still in progress. 

So I paused my hook for B.Create a moment, then the transaction of A will be complete, so that my hook will continue to work.

How can I know that the other pending transaction is on when a custom hook is triggered?

**Reply by rijkvanzanten** [reply]
> How can I know that the other pending transaction is on when a custom hook is triggered?

I don't think there's a way to actively wait for the transaction to finish. This is also being discussed and tracked as a bug over on https://github.com/directus/directus/issues/13508
